assets/js/directives/featureSliceVisualizer.coffee

```diff
@@ -64,33 +65,24 @@ app.directive 'featureSliceVisualizer', [
               data: []
           return
 
-        retrieveSamples = (feature) ->
-          if not feature.samples?
-            return backendService.retrieveSamples feature.id
-              .then (samples) -> feature.samples = samples
-              .fail console.error
-          else
-            return $q.resolve feature.samples
-
         retrieveMarginalDistribution = ->
+          if scope.targetFeature.marginalProbDistr
+            return $q.resolve scope.targetFeature.marginalProbDistr
           return backendService.getExperiment()
             .then (experiment) -> experiment.getProbabilityDistribution []
-            .then (distr) -> scope.targetFeature.marginalProbDistr = distr
+            .then (response) ->
+              scope.targetFeature.marginalProbDistr = response.distribution
             .fail console.error
 
         scopeUtils.waitForVariableSet scope, 'targetFeature'
           .then ->
-
             # Set defaults for xFeature and yFeature
             if scope.selectedFeatures.length >= 2
               scope.xFeature = scope.selectedFeatures[0]
               scope.yFeature = scope.selectedFeatures[1]
             
-            promises = scope.selectedFeatures.map retrieveSamples
-            promises.push retrieveSamples scope.targetFeature
-            promises.push retrieveMarginalDistribution()
-            $q.all promises
-              # TODO remove this annoying timeout
+            retrieveMarginalDistribution()
+              # TODO remove timeout, it waits until layouting is done
               .then -> $timeout null, 1500
               .then -> scope.setupCharts()
               .then -> scope.updateCharts()
```

Comment:
You're right.
But that would call SetupCharts with the timeout cancel fnc and updateChart with the return value of setupCharts.
This is more explicit and less prone to errors I think.